include/wally_descriptor.h

```diff
@@ -0,0 +1,145 @@
+#ifndef LIBWALLY_CORE_DESCRIPTOR_H
+#define LIBWALLY_CORE_DESCRIPTOR_H
+
+#include "wally_core.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+struct wally_map;
+
+/* Miniscript type flag */
+#define WALLY_MINISCRIPT_WITNESS_SCRIPT 0x00 /** Witness script */
+#define WALLY_MINISCRIPT_TAPSCRIPT      0x01 /** Tapscript */
+
+/**
+ * Canonicalize a descriptor.
+ *
+ * :param descriptor: Output descriptor.
+ * :param vars_in: Map of variable names to values, or NULL.
+ * :param flags: For future use. Must be 0.
+ * :param output: Destination for the resulting canonical descriptor.
+ *|    The string returned should be freed using `wally_free_string`.
+ *
+ * Replaces any variables from ``vars_in`` with their mapped values,
+ * and adds a checksum if required. Key names for ``vars_in`` must be 16
+ * characters or less and start with a letter.
+ *
+ * .. note:: Other canonicalization (hardened derivation indicator
+ * mapping, and private to public key mapping) is not yet implemented.
+ */
+WALLY_CORE_API int wally_descriptor_canonicalize(
```

Comment:
Many of these checks would require full parsing of the descriptor, which is the responsibilitiy of the other functions. I originally intended this call to allow full canonicalisation of the input (for example, sortedmulti with x,y,z is the same as x,z,y or other permutations), however it turns out that core descriptors have no such concept, and do not compute checksums on a canonical representation which puts limits on the kinds of cleanup that are useful in practice, and the ability to compare descriptors when they can be trivially permuted.

However the primary reason I moved this back to WIP status is the script generation, which will also need interface changes, so I will re-review this function as part of that.

Reply:
You could also add a flag to perform (or skip) full validation, or have another method for it. But in practice both validating and getting the canonical version are something any safe software should probably do on data from the outside. Otherwise software has to delay error handling until some later step.

While you're changing things: can you add a flag to mandate a checksum, for all functions? I found myself parsing the string for `#`. The current code already ensures that _if_ a checksum is present it's correct, but it doesn't care if a checksum is missing.

Reply:
>  I found myself parsing the string for #.

You can canonicalize the string and then compare it to your input, if canonicalization fails the input is invalid, if it succeeds and the result differs then the original had no checksum.

I'll consider adding a flag to validate.

Reply:
Doesn't canonicalization change `h` to ` and thus change the checksum?

Reply:
>Doesn't canonicalization change h to ` and thus potentially change the checksum?

No, all canonicalization currently does is variable substitution and append a checksum if missing.

Reply:
updated to a parse-then-query API so resolving this